config_utilities/include/config_utilities/factory.h

```diff
@@ -35,184 +35,269 @@
+  virtual ~ConfigWrapper() = default;
+  virtual std::unique_ptr<ConfigWrapper> clone() const = 0;
+  virtual void onDeclareConfig() = 0;
+  virtual std::unique_ptr<ConfigWrapper> createDefault() const = 0;
+  std::string type;
+};
+
+struct FactoryMapBase {
+  virtual ~FactoryMapBase() = default;
+  virtual bool hasEntry(const std::string& type) = 0;
+  virtual void removeEntry(const std::string& type) = 0;
+  virtual bool empty() = 0;
+};
+
+//! @brief Struct to store the factory methods for the creation of modules.
+template <typename Factory>
+struct FactoryMap : FactoryMapBase {
+  //! @brief Add entries to the map with verbose warnings.
+  bool addEntry(const std::string& type, const Factory& method) {
+    const auto has_type = map.find(type) != map.end();
+    if (has_type) {
+      return false;
     }
-    ss << "\n}";
-    return ss.str();
+
+    map.emplace(type, method);
+    return true;
   }
 
- private:
-  static ModuleRegistry& instance() {
-    static ModuleRegistry instance_;
-    return instance_;
+  //! @brief Get factory for type (returns an invalid function if type not registered)
+  Factory getEntry(const std::string& type) {
+    auto iter = map.find(type);
+    return iter == map.end() ? Factory() : iter->second;
   }
 
-  ModuleRegistry() = default;
+  bool hasEntry(const std::string& type) override { return map.find(type) != map.end(); }
+  void removeEntry(const std::string& type) override { map.erase(type); }
+  bool empty() override { return map.empty(); }
 
-  // Nested modules: base_type -> args -> registered <type_name, tpye>.
-  std::map<std::string, std::map<std::string, std::vector<std::pair<std::string, std::string>>>> modules;
+ private:
+  std::map<std::string, Factory> map;
 };
 
-// Struct to store the factory methods for the creation of modules.
-template <typename FactoryMethod>
-struct ModuleMapBase {
-  using FactoryMethodMap = std::unordered_map<std::string, FactoryMethod>;
-
-  // Add entries to the map with verbose warnings.
-  static bool addEntry(const std::string& type, const FactoryMethod& method, const std::string& type_info) {
-    FactoryMethodMap& map = instance().map;
-    if (map.find(type) != map.end()) {
-      if (!type_info.empty()) {
-        Logger::logError("Cannot register already existent type '" + type + "' for " + type_info + ".");
+class ModuleRegistry {
+ public:
+  using LockCallback = std::function<void(const ModuleInfo&, std::string, std::string)>;
+  using CreateCallback = std::function<void(const ModuleInfo&, std::string, void*)>;
+
+  template <typename BaseT, typename DerivedT, typename... Args>
+  static bool addModule(const std::string& type, FactoryMethod<BaseT, Args...> method, bool skip_first_arg = false) {
+    const auto key = ModuleInfo::fromTypes<BaseT, Args...>(skip_first_arg);
+    using Factory = FactoryMethod<BaseT, Args...>;
+    if (locked()) {
+      if (hasModule(key, type)) {
+        return false;
```

Comment:
Intentionally silently fails (I'll add a note). The check is to avoid compiled-in factories that get re-registered in the external library (e.g., `StdoutLogger`)